### PR TITLE
v3.12.1

### PR DESCRIPTION
**Pull request by schneems**

### CODE REVIEW

**Comment by MSP-Greg** [comment]
@schneems

Please consider releasing the pre-compiled gems (with SSL) for windows/mingw.

Thanks for all your work with Puma, Greg

**Comment by schneems** [comment]
Hey @MSP-Greg. I'm having some issues with my local install of jruby. I need to get that figured out to release a jruby gem, so we have a little time.

Can you give me details on how to grab and push the binaries for windows? It's not a thing I have much experience with.

As a related side note, not providing vendored versions actually has allowed teams (unintentionally) to vendor gems and still deploy to Heroku. As long as none of the gems are pre-compiled, then it doesn't matter if you're installing from windows/mac and deploying to linux.

**Comment by MSP-Greg** [comment]
@schneems 

No rush.  One thing also is the Ruby 2.6.0 issue.  I'm not sure if it's a Ruby issue, or an issue that can be dealt with in Puma.  Because it only occurs when using `fork`, I can't test it with Windows...

> Can you give me details on how to grab and push the binaries for windows? It's not a thing I have much experience with.

I also haven't pushed any actual gems for quite a while.  Go to:

https://ci.appveyor.com/project/puma/puma

There you'll see the builds for 64 bit and 32 bit.  Click on either, and you'll go to the default log page.  Then, click on 'artifacts', and you'll see the gem file available for download.  Rather than setting up Appveyor, I assume you'll download them and push manually?

BTW, feel free to have any Windows types give them a try, as they are saved with every commit/pr...

> not providing vendored versions actually has allowed teams (unintentionally) to vendor gems and still deploy to Heroku

I assume there's some way to specify something like "--platform=ruby"?  Not sure how deploy works, but Bundler & RubyGems should never install a gem on an improper OS, they should default to the 'Ruby' (non-compiled) gem?

For non-windows users, most don't have to even consider pre-compiled gems, as none match their OS/platform.  My normal ruby version is trunk, and there are several ways to force non-compiled gems to install, as very few are compiled to run on trunk...

#### Why not let windows users compile Puma?

I don't use *nix or MacOS locally, hence, most of what I know is from working on Travis.  For non-Windows use, it's my impression that OpenSSL is typically defined by the OS.

Since OpenSSL isn't 'native' to Windows, the version used is defined by a combination of the packager's decision, what the Ruby version works with, and what OpenSSL packages are available for the build system.  Building the gem for Ruby 2.2 thru 2.6/2.7 is using two build systems and three OpenSSL packages.

Both here and at EventMachine, there are always issues about "I can't build on Windows with OpenSSL", as some of the users are not experienced with build systems and packages.  Personally, I'd like to hope that at some point they'd want some experience with them, but I don't think that should stop them from learning Ruby/Rails/Puma, etc.

File: History.md [hunk 1]
```diff
@@ -1,3 +1,18 @@
...
+* x bugfixes
+
+## 3.12.1 / 2019-01-08
+
+* 1 features
+  * Internal strings are frozen (#1649)
+* 3 bugfixes
+  * Fix chunked ending check (#1607)
+  * Rack handler should use provided default host (#1700)
+  * Better support for detecthing runtimes that support `fork` (#1630)
```
Comment (olleolleolle) [inline History.md]:
```suggestion
  * Better support for detecting runtimes that support `fork` (#1630)
```

**Comment by harmdewit** [comment]
Any idea when this get released? 🙂

**Comment by olleolleolle** [comment]
Here's the [current compare to master](https://github.com/puma/puma/compare/f7e1e58dcba1e19e470b6eb0c44c33a0722285f6...master)

**Comment by schneems** [comment]
Tried again today to release, and I'm still getting jruby errors. Not sure what's going on

```
$ be rake java gem
javac -extdirs "/Users/rschneeman/Library/Java/Extensions:/Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/lib/ext:/Library/Java/Extensions:/Network/Library/Java/Extensions:/System/Library/Java/Extensions:/usr/lib/java" -target 1.6 -source 1.6 -Xlint:unchecked  -cp "/Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/lib/resources.jar:/Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/lib/rt.jar:/Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/lib/sunrsasign.jar:/Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/lib/jsse.jar:/Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/lib/jce.jar:/Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/lib/charsets.jar:/Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/lib/jfr.jar:/Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/classes:/Users/rschneeman/.rubies/jruby-9.2.6.0/lib/jruby.jar:" -d tmp/java/puma_http11 ext/puma_http11/PumaHttp11Service.java ext/puma_http11/org/jruby/puma/Http11.java ext/puma_http11/org/jruby/puma/Http11Parser.java ext/puma_http11/org/jruby/puma/MiniSSL.java
ext/puma_http11/PumaHttp11Service.java:5: cannot access org.jruby.Ruby
bad class file: org/jruby/Ruby.class(org/jruby:Ruby.class)
class file has wrong version 52.0, should be 50.0
Please remove or make sure it appears in the correct subdirectory of the classpath.
import org.jruby.Ruby;
                ^
rake aborted!
Command failed with status (1): [javac -extdirs "/Users/rschneeman/Library/...]
/Users/rschneeman/.gem/jruby/2.5.3/gems/rake-compiler-1.0.7/lib/rake/javaextensiontask.rb:98:in `block in define_compile_tasks'
```

**Comment by MSP-Greg** [comment]
Maybe have a look at issues:

[1725 Errno::ECHILD: No child processes](https://github.com/puma/puma/issues/1725)

and

[1731 uninitialized constant Puma::Cluster](https://github.com/puma/puma/issues/1731)

**Comment by schneems** [comment]
Figured out the issue, or at least worked around it with a reinstall of java via homebrew cask. Released 3.12.1.

**Comment by schneems** [comment]
Thanks @MSP-Greg I will take a look, btw are you going to be going to any Ruby Conferences this year? RailsConf or Kaigi or Euruko?

**Comment by MSP-Greg** [comment]
You're welcome.  As to Ruby conferences, at present, no.  If I do (or next year), I'll let you know.  Thanks for asking...

**Comment by headius** [comment]
👍 Yay, thanks for getting a release out!

**Comment by MSP-Greg** [comment]
@schneems

I assume it was decided not to release pre-compiled Windows gems, which is ok.

**Pull request merged**